Fix get and set of table Row when using multiple column names

taldcroft opened the pull request:
Fixes #7049.

For now I milestoned this for 3.1 because there is no 3.0.1 in master CHANGES.rst.  

From a functional perspective I think this could be backported to LTS, but it is slightly a pain because I factored out a new helper method and it has now a `isinstance(.., str)` instead of `isinstance(.., six.string_types)`.

taldcroft commented:
@eteq @adrn - there is a weird failure in test_transform_to_nonscalar_nodata_frame() for appveyor, which should be unrelated to this PR.  Thoughts?

taldcroft commented:
Nevermind, presumably my test fail is related to #7105.

mhvk commented on astropy/table/table.py:
```diff
@@ -1441,6 +1436,18 @@ def dtype(self):
+                all(isinstance(x, str) for x in names)):
+            bad_names = [x for x in names if x not in self.colnames]
+            if bad_names:
+                raise ValueError('Slice name(s) {0} not valid column name(s)'
```
I guess this is too late, but shouldn't this be `KeyError`? That is what one gets with a non-existing single column name. Note that if one were to just remove this test, it would give the `KeyError` later anyway.

taldcroft replied:
Yup, `KeyError` would make more sense.  Sigh.

mhvk commented on astropy/table/row.py:
```diff
@@ -38,10 +38,24 @@ def __init__(self, table, index):
     def __setitem__(self, item, val):
-        self._table.columns[item][self._index] = val
+        if self._table._is_list_or_tuple_of_colnames(item):
+            if len(val) != len(item):
```
I'm not sure I see the point of this check: if someone puts in a scalar, they now get a error that is not all that obvious, and it excludes iterables without a length. Maybe best to just over the relevant part of `Table.__setitem__` so things are consistent. Alternatively, maybe we pass this back to `Table`, by just doing
```
self._table[item][self._index] = val
```
But that perhaps has more overhead than warranted.

taldcroft replied:
Interesting point which has a lot of implications:
- I had not considered the idea of essentially broadcasting a scalar.  This is not supported in `insert_row`, and it seems like an unusual thing to do.  Agreed that catching this more nicely would be good.
- Likewise the iterable for row values seems obscure.  It is certainly important that the supplied values exactly match in length the number of columns.  I guess iterables could be supported by checking for an iterable and doing `tuple(..)` around it?
- The goal (admittedly not entirely achieved) is to make setting the row transactional, so if it fails half way you don't half-way update the table.  `insert_row` does do this.  The existing code here catches two common errors up front, namely bad column name and mismatch between columns and values.  What it doesn't catch is where setting fails (`self._table.columns[name][self._index] = val_`).  So that should be fixed.

Anyway, point taken, I'm thinking about it.

taldcroft replied:
How about this for the first two bullets, which basically says (nicely) "don't do that".  I think it is not an unreasonable restriction that the right side is a sequence (with a length) and that it match.
```
            try:
                assert len(val) == len(item)
            except Exception:
                raise ValueError('right hand side must be a sequence of values with '
                                 'the same length as the number of selected columns')
```

mhvk replied:
Yes, this does the same as what you have in table but is cleaner. Good enough for now!

taldcroft commented on astropy/table/row.py:
```diff
@@ -38,10 +38,24 @@ def __init__(self, table, index):
     def __getitem__(self, item):
-        return self._table.columns[item][self._index]
+        if self._table._is_list_or_tuple_of_colnames(item):
+            cols = [self._table[name] for name in item]
```
Sadly, table column slicing (`t['col0', 'col1', ..]`) returns a *copy*, and this does seem like  a mistake. In any case the slice here definitely needs to be a reference.

Using the `columns` dict directly will help performance (slightly), and this is for 3.1!  :smile:

mhvk replied:
OK, I hadn't realised that by default you get a copy -- that indeed seems like a mistake.
But better addressed separately.

taldcroft commented:
@mhvk - I reworked this a fair bit based on your comments and I think it is better now.  I added rollback of the row update if anything goes wrong, which then removed the need for up-front testing of the column names.  Along the way I discovered that `t[1] = values` was not transactional and could do a halfway update, but now this is and Row.__setitem__ are handled with the same code.  Yay.

There is an API change in the exception that gets raised, as seen in test changes.  Probably OK?

astrofrog commented:
> There is an API change in the exception that gets raised, as seen in test changes. Probably OK?

Yes, but please mention it in the changelog

mhvk commented on astropy/table/table.py:
```diff
@@ -2165,6 +2157,27 @@ def rename_column(self, name, new_name):
         self.columns[name].info.name = new_name
 
+    def _set_row(self, idx, colnames, vals):
+        try:
```
It is probably overkill, but we could be even more agnostic about `value` by replacing this with, just after the loop and before the exception:
```
# Check that we got to the end of names (which, unlike vals,
# is known to be a list of tuple of str, so indexable).
if name is not names[-1]:
    raise ValueError('right ...')
```
It should even be slightly faster... But perhaps not as clear (and clarity does matter).

taldcroft replied:
I don't think that catches the case where `vals` is an iterable that is *longer* than `colnames` (where `zip` cuts off because of `colnames`).

mhvk replied:
true, well, that decides it I think!

taldcroft commented:
Just realized with an API change this should not really be in a bug fix release (3.0.1).  What do you think?

mhvk commented:
Can we just put it in 3.0 still? It does fix a bug.

bsipocz commented:
Hmm. We have rc2 out, but I'm inclined that this is still fine as a bugfix to get into the final 3.0 release. I would be happier to do that than 3.0.1, as it would also solve the KeyError API change dilemma.
Any veto for that @eteq or @astrofrog (or anyone else)?

taldcroft commented:
Assuming that others agree we can get this into 3.0, I have updated the change log accordingly (and please review since I added a bunch).  Also changed the milestone and squashed to one commit.

taldcroft commented:
BTW, not to shoot myself in the foot, I'll say that I am somewhat ambivalent about pushing this into 3.0 without a proper RC, esp. since this involves an API change.

taldcroft commented:
But of course it's up to release managers.  :smile:

bsipocz commented:
Ouch. Well, I assumed it'll be fine as the change in the type of exception is rather a minor API change, but your previous comment made me ambivalent, too. 
But, if you think it shouldn't go into 3.0 without RC testing, then it cannot really go into 3.0.1 either for the same API change reason.

astrofrog commented:
Can we not decouple the exception type change from the rest of the fixes?

mhvk commented:
I find it *very* hard to believe the type of exception raised for multiple column names was relied on anywhere, while this does fix a rather annoying bug. Since I'm also not that sure `rc` gets as much testing as a proper release, which means in a way 3.0.0 is the true "release candidate", I'm +1 for going for 3.0.0.

taldcroft commented:
It is not easy to decouple the exception change because of the way the code logic is updated and cleaned up.

I do agree with @mhvk that this is very unlikely to be a problem.  It's just that we did get burned once on a late addition to the release and made a resolution to not do that again.  Hence the ambivalence.  

I *guess* one way to look at this is say the only utility in making rc3 and putting it out for a week is to give downstream codes an extra week to get fixed, or complain.  Very worst case (highly unlikely) is a quick 3.0.1 that addresses a problem.

bsipocz commented:
I'm fine with doing a quick 3.0.1 if needed. Also, many of the affiliates tests against the dev version, we may just check those out to be sure before the backport.

astrofrog commented:
@taldcroft - can you rebase to fix the changelog conflict?

taldcroft commented:
This bug is showing up right now in some Chandra operations analysis I'm doing.  😝
